File: src/main/java/com/codeclassic/grubby/api/dto/GenerateBrdRequest.java
package com.codeclassic.grubby.api.dto;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Pattern;
import jakarta.validation.constraints.Size;
import lombok.Data;

import java.util.Map;

/**
 * P2 — repoUrl now carries basic format constraints (scheme + length).
 * Deep SSRF validation (IP range + DB allowlist) happens in RepoUrlValidator,
 * called from BrdOrchestratorService.submit() before any clone is attempted.
 */
@Data
public class GenerateBrdRequest {

    /**
     * Must start with an allowed scheme.  Full SSRF validation is done server-side
     * in RepoUrlValidator — this annotation is a first-pass sanity check only.
     */
    @NotBlank
    @Size(max = 500, message = "repoUrl must not exceed 500 characters")
    @Pattern(
        regexp = "^(https?|git|ssh)://.*",
        message = "repoUrl must start with https://, http://, git://, or ssh://"
    )
    private String repoUrl;

    @Size(max = 200)
    private String branch;

    @Size(max = 40)
    private String commitSha;

    @Size(max = 5000)
    private String featureContext;

    @Size(max = 100)
    private String projectType;

    /** none | token | oauth */
    @Size(max = 20)
    private String authType;

    /** Git auth token — never persisted beyond the active job */
    @Size(max = 500)
    private String authToken;

    private boolean forceReanalyze;

    /** Model identifier, e.g. "gpt-4o-mini" or "claude-opus-4-7". Null/blank = server default. */
    @Size(max = 100)
    private String aiModel;

    /** Optional Slack Incoming Webhook URL for completion notifications. */
    @Size(max = 500)
    private String slackWebhookUrl;

    private Map<String, String> options;
}

File: grubbyflut/lib/models/generate_brd_request.dart
class GenerateBrdRequest {
  GenerateBrdRequest({
    required this.repoUrl,
    this.branch,
    this.commitSha,
    this.featureContext,
    this.projectType,
    this.authType,
    this.authToken,
    this.forceReanalyze = false,
    this.options = const {},
  });

  String repoUrl;
  String? branch;
  String? commitSha;
  String? featureContext;
  String? projectType;
  String? authType; // none|token|oauth
  String? authToken; // only for dev; in prod use vault approach
  bool forceReanalyze;
  Map<String, String> options;

  Map<String, dynamic> toJson() => {
        'repoUrl': repoUrl,
        if (branch != null && branch!.isNotEmpty) 'branch': branch,
        if (commitSha != null && commitSha!.isNotEmpty) 'commitSha': commitSha,
        if (featureContext != null && featureContext!.isNotEmpty) 'featureContext': featureContext,
        if (projectType != null && projectType!.isNotEmpty) 'projectType': projectType,
        if (authType != null && authType!.isNotEmpty) 'authType': authType,
        if (authToken != null && authToken!.isNotEmpty) 'authToken': authToken,
        'forceReanalyze': forceReanalyze,
        if (options.isNotEmpty) 'options': options,
      };
}
